projects/rulesets-base/src/extended-rules/spectral-rule.ts

```diff
@@ -0,0 +1,107 @@
+import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
+import {
+  IFact,
+  IChange,
+  OpenAPIV3,
+  Result,
+  constructFactTree,
+  getFactForJsonPath,
+} from '@useoptic/openapi-utilities';
+import { ExternalRuleBase } from '../rules/external-rule-base';
+
+type Lifecycle = 'added' | 'addedOrChanged' | 'changed' | 'always';
+
+// Spectral results will be returned if they fail
+export interface SpectralResult {
+  path: (string | number)[];
+  code: string | number;
+  message: string;
+}
+
+export interface Spectral {
+  run: (jsonSpec: OpenAPIV3.Document) => Promise<SpectralResult[]>;
+}
+
+function toOpticResult(
+  spectralResult: SpectralResult,
+  lifecycle: Lifecycle,
+  relevantChangeOrFact: IFact | IChange
+): Result {
+  return {
+    condition: spectralResult.code.toString(),
+    passed: false,
+    error: spectralResult.message,
+    isMust: true,
+    isShould: false,
+    where: `${lifecycle} `,
+    change: relevantChangeOrFact,
+  };
+}
+
+export class SpectralRule extends ExternalRuleBase {
+  private lifecycle: Lifecycle;
+  constructor(
+    private spectral: Spectral,
+    options?: {
+      applies?: Lifecycle;
+    }
+  ) {
+    super();
+    this.lifecycle = options?.applies ?? 'always';
+  }
+
+  async runRules(inputs: {
+    context: any;
+    nextFacts: IFact[];
+    currentFacts: IFact[];
+    changelog: IChange[];
+    nextJsonLike: OpenAPIV3.Document<{}>;
+    currentJsonLike: OpenAPIV3.Document<{}>;
+  }): Promise<Result[]> {
+    const factTree = constructFactTree(inputs.nextFacts);
+
+    const changesByJsonPath: Record<string, IChange> = inputs.changelog.reduce(
+      (acc, next) => {
+        acc[next.location.jsonPath] = next;
+        return acc;
+      },
+      {}
+    );
+
+    const spectralResults = await this.spectral.run(inputs.nextJsonLike);
+    const results: Result[] = [];
+
+    for (const spectralResult of spectralResults) {
+      const path = jsonPointerHelpers.compile(
+        spectralResult.path.map((p) => String(p))
+      );
+      const fact = getFactForJsonPath(path, factTree);
```

Comment:
95% of spectral rules will point at attributes, not what we call facts. Ie they point at `operationId` not the operation

Is this doing strict equality? child of?

Reply:
It's finding the closest parent fact from a json path.

E.g. `/api~1users/get/operationId` will return the Fact at `/api~1users/get`

Reply:
Awesome -- it looked like it might, but wasn't sure w/o running it.